jung-api/src/test/java/edu/uci/ics/jung/graph/CTreeMutationTest.java

```diff
@@ -229,12 +233,10 @@ private int controlledNumEdgesToRemove(Random gen) {
 
   private static <N> List<EndpointPair<N>> descendantEdges(
       CTree<N> tree, EndpointPair<N> startingEdge) {
-    // TODO: Migrate to com.google.common.graph.Traverser when it is available.
-    TreeTraverser<EndpointPair<N>> traverser = TreeTraverser.using(edge -> outEdges(tree, edge));
-    return traverser
-        .breadthFirstTraversal(startingEdge)
+    Traverser<EndpointPair<N>> traverser = Traverser.forTree(edge -> outEdges(tree, edge));
+    return Streams.stream(traverser.breadthFirst(startingEdge))
```

Comment:
I should have taken a closer look at this method before commenting originally, sorry.  Let's back up a bit.

(1) Unless I'm missing something, descendantEdges() doesn't need to return a List.  Collections.shuffle() is called on it, but I don't see any reason why that's necessary; the only thing it affects is how those elements appear in edgesRemoved2, which I don't think makes any functional difference to your tests or their coverage.  
Am I missing something?

(2) There's generally no particular reason to create a separate `Traverser` object; you can (and probably should) just call `breadthFirst()` directly:

    Traverser.forTree(...).breadthFirst(...);

(3) If descendantEdges() can return an Iterable, then you can just do this:

    return Iterables.skip(Traverser.forTree(...).breadthFirst(...), 1);

(4) It seems a bit odd to do the transformation from nodes to edges inside the Traverser; why not do that at the last minute, inside the calling method?

Thus:

    // start traversing from leafEdge.source() so you don't need to skip the first element
    Iterable<Integer> descendants = Traverser.forTree(tree).breadthFirstTraversal(leafEdge.source());
    ...
    for (Integer descendant : descendants) {
      // not 100% sure if you want the parent or each child; don't assume this is correct :D
      Integer parent = tree.predecessor(descendant).get();  
      assertThat(tree.edges.contains(EndpointPair.ordered(parent, descendant))).isFalse();
    }

    ...
    Iterables.addAll(edgesRemoved2, predecessors);

This way you don't need the descendantEdges() or outEdges() methods at all.

WDYT?

Reply:
Very good questions! I'll do my best to answer them in order. :)

(1) I'm not 100% sure that I understand what you're saying here. I understand you're saying that the type could be turned to, say, `Iterable`, since we don't really need to shuffle elements, as doing so doesn't buy us anything in this case. And I think I agree. (I must've blindly followed Guava's `AbstractGraphTest` here, as it was also shuffling elements for some reason.) :)

But FYI, I've decided to remove `descendantEdges()` and replace it with a `subTree()` method, which I think makes the code conceptually easier to digest. I've submitted commit https://github.com/jrtom/jung/pull/206/commits/8f5325626c7c3a2c6b03043b3d70effe1b04adfc to show what I mean.

(2) Indeed, there is in principle no reason to create a separate `Traverser` object. But for my installed JDK - Oracle, Java 8, update 161 - the compiler wasn't advanced enough to determine the generics for

```java
    return Streams.stream(
            Traverser.forTree(edge -> outEdges(tree, edge)).breadthFirst(startingEdge))
        .skip(1) // skip the starting edge itself
        .collect(toCollection(ArrayList::new));
```

It explicitly needed either:

```java
    return Streams.stream(
            Traverser.<EndpointPair<N>>forTree(edge -> outEdges(tree, edge))
                .breadthFirst(startingEdge))
        .skip(1) // skip the starting edge itself
        .collect(toCollection(ArrayList::new));
```

or the following (which was the existing solution):

```java
    Traverser<EndpointPair<N>> traverser = Traverser.forTree(edge -> outEdges(tree, edge));
    return Streams.stream(traverser.breadthFirst(startingEdge))
        .skip(1) // skip the starting edge itself
        .collect(toCollection(ArrayList::new));
```

and I thought that the latter option was the easiest to read.

But I think this is all a bit moot now, because in commit https://github.com/jrtom/jung/pull/206/commits/8f5325626c7c3a2c6b03043b3d70effe1b04adfc I got rid of `descendantEdges()`. :)

(3) Probably N/A. :)

(4) I found it difficult to understand from your example code what you wanted me to do. However, in commit https://github.com/jrtom/jung/pull/206/commits/8f5325626c7c3a2c6b03043b3d70effe1b04adfc , I think I've managed to simplify things on a conceptual level (that is, remove the sub-tree rooted at `leafEdge.source()`, instead of some list of edges), which I hope achieves what you were looking for here.

(It also has the side effect of getting rid of `descendantEdges()` and `outEdges()`, so yay!)